pkg/gatewaysecret/handler.go

```diff
@@ -161,21 +170,44 @@ func GetGatewaySecret(ctx context.Context, clnt client.Client) (*apicorev1.Secre
 	return secret, nil
 }
 
-func (gsh *GatewaySecretHandler) StartRootCertificateWatch(clientset *kubernetes.Clientset,
-	log logr.Logger,
-) {
+func (h *Handler) StartRootCertificateWatch() {
 	ctx, cancel := context.WithCancel(context.TODO())
 	defer cancel()
 
-	secretWatch, err := clientset.CoreV1().Secrets(istioNamespace).Watch(ctx, apimetav1.ListOptions{
+	h.handleAlreadyCreatedRootCertificate(ctx)
+	h.handleNewRootCertificates(ctx)
+}
+
+func (h *Handler) handleAlreadyCreatedRootCertificate(ctx context.Context) {
+	rootCASecrets, err := h.kcpClientset.CoreV1().Secrets(istioNamespace).List(ctx, apimetav1.ListOptions{
+		FieldSelector: fields.OneTermEqualSelector(apimetav1.ObjectNameField, kcpRootSecretName).String(),
+	})
+	if err != nil {
+		h.log.Error(err, "unable to list root certificate")
+		panic(err)
+	}
+	if len(rootCASecrets.Items) != 1 {
+		h.log.Error(errExpectedOneRootCASecret, errExpectedOneRootCASecret.Error(),
+			"found", len(rootCASecrets.Items))
+		panic(fmt.Errorf("%w: found %d", errExpectedOneRootCASecret, len(rootCASecrets.Items)))
+	}
+	rootCASecret := &rootCASecrets.Items[0]
+	err = h.manageGatewaySecret(ctx, rootCASecret)
+	if err != nil {
+		panic(err)
+	}
+}
+
+func (h *Handler) handleNewRootCertificates(ctx context.Context) {
+	secretWatch, err := h.kcpClientset.CoreV1().Secrets(istioNamespace).Watch(ctx, apimetav1.ListOptions{
 		FieldSelector: fields.OneTermEqualSelector(apimetav1.ObjectNameField, kcpRootSecretName).String(),
 	})
 	if err != nil {
-		log.Error(err, "unable to start watching root certificate")
+		h.log.Error(err, "unable to start watching root certificate")
 		panic(err)
```

Comment:
if you panic in goruntine, it just crashed this goruntine itself, but in your case, if can't watching root certificate, the whole lifecycle manager should not started, right?

In this case, you need to send error back to go main goruntine and panic there. So you should use channel here.

Another option is not design whole `StartRootCertificateWatch` as goruntine, instead, before `WatchEvents`, you can still running in the main goruntine, you only need to make `WatchEvents` as a dedicated thread, it can simplify your panic handling.